src/webviews/extension-side/dataviewer/kernelDataViewerDependencyImplementation.ts

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+'use strict';
+
+import { SemVer } from 'semver';
+import { ProductNames } from '../../../kernels/installer/productNames';
+import { Product } from '../../../kernels/installer/types';
+import { IApplicationShell } from '../../../platform/common/application/types';
+import { traceWarning } from '../../../platform/logging';
+import { DataScience, Common } from '../../../platform/common/utils/localize';
+import { EnvironmentType } from '../../../platform/pythonEnvironments/info';
+import { sendTelemetryEvent, Telemetry } from '../../../telemetry';
+import { executeSilently } from '../../../kernels/helpers';
+import { IKernel } from '../../../kernels/types';
+import { parseSemVer } from '../../../platform/common/utils';
+import { IDataViewerDependencyService } from './types';
+import { pandasMinimumVersionSupported } from './constants';
+
+export const kernelGetPandasVersion =
+    'import pandas as _VSCODE_pandas;print(_VSCODE_pandas.__version__);del _VSCODE_pandas';
+
+function kernelPackaging(kernel: IKernel): '%conda' | '%pip' {
+    const envType = kernel.kernelConnectionMetadata.interpreter?.envType;
+    const isConda = envType === EnvironmentType.Conda;
+    // From https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-pip (%conda is here as well).
+    return isConda ? '%conda' : '%pip';
+}
+
+/**
+ * Uses the Kernel to manage the dependencies of a Data Viewer.
+ */
+export class KernelDataViewerDependencyImplementation implements IDataViewerDependencyService {
+    constructor(private readonly applicationShell: IApplicationShell, private isCodeSpace: boolean) {}
+
+    protected async execute(command: string, kernel: IKernel): Promise<(string | undefined)[]> {
+        const outputs = await executeSilently(kernel.session!, command);
```

Comment:
The assumption that the kernel has a session (with the !) seems like it might be sneaking around the type system a bit. I don't see, at least at this point, an explicit reason the session would be undefined. Per the usage of this function it look like it should maybe just take a Session as a non optional parameter? I don't believe anything else on the Kernel is used, and the caller should guarantee that there is a session.

Reply:
Yeah, I see the check at the start of the dependency check later in the code. Seems like after that you could just use the session.

Reply:
This function is called by a function that checks for the session, but I like your idea! Passing the session directly, rather than the kernel. Thank you 🙏

Reply:
How about this? https://github.com/microsoft/vscode-jupyter/pull/10825/commits/c3341d788e63dff03d4f124a89bf57dfc72c592a